app/client-v2/src/stores/projectStore.ts

```diff
@@ -75,50 +76,76 @@ export const useProjectStore = defineStore("project", {
-      for (const file of files) {
-        const content = await file.text();
-        const fileHash = Md5.hashStr(content);
-        this.project.samples.push({ hash: fileHash, filename: file.name });
+      this.uploadingPercentage = 0;
+      const hashedFileBatches = await this.batchFilesForProcessing(files);
+      this.processFileBatches(hashedFileBatches);
+    },
+    async batchFilesForProcessing(files: File[]) {
+      const batchSize = 5; // found to be best balance of throughput and memory usage
+      const hashedFiles: HashedFile[] = await Promise.all(
+        files.map(async (file) => {
+          const content = await file.text();
+          const fileHash = Md5.hashStr(content);
+          return { hash: fileHash, filename: file.name, file };
+        })
+      );
+      const hashedFileBatches: HashedFile[][] = [];
+      for (let i = 0; i < hashedFiles.length; i += batchSize) {
+        hashedFileBatches.push(hashedFiles.slice(i, i + batchSize));
+      }
+      return hashedFileBatches;
+    },
+    async processFileBatches(hashedFileBatches: HashedFile[][]) {
+      const maxWorkers = this.getOptimalWorkerCount();
+      const activeBatches: Set<Promise<void>> = new Set();
+      let uploadPercentNumerator = 0;
 
-        // run web worker to get sketch and amr data and then post to server
+      for (const hashedFileBatch of hashedFileBatches) {
+        if (activeBatches.size >= maxWorkers) {
+          await Promise.race(activeBatches); // wait for at least 1 batch to finish
+        }
+        const batchPromise = this.computeAmrAndSketch(hashedFileBatch);
+        activeBatches.add(batchPromise);
+
+        batchPromise
+          .catch(() => console.log("error processing batch"))
+          .finally(() => {
+            activeBatches.delete(batchPromise);
+            uploadPercentNumerator++;
+            this.uploadingPercentage = Math.round((uploadPercentNumerator / hashedFileBatches.length) * 100);
+          });
+      }
+    },
+    getOptimalWorkerCount() {
+      return Math.floor(navigator.hardwareConcurrency * 0.5) || 4;
+    },
+    async computeAmrAndSketch(hashedFiles: HashedFile[]): Promise<void> {
+      return new Promise((resolve, reject) => {
         const worker = new Worker("/worker.js");
-        worker.postMessage({ hash: fileHash, fileObject: file });
+        worker.postMessage(hashedFiles);
 
-        worker.onmessage = async (event: MessageEvent<WorkerResponse>) => {
-          this.handleWorkerResponse(file.name, event);
+        worker.onmessage = async (event: MessageEvent<WorkerResponse[]>) => {
+          this.handleWorkerResponse(event.data);
+          worker.terminate();
+          resolve();
         };
+
         worker.onerror = (error) => {
           console.error(error);
-          this.toast.showErrorToast("Ensure uploaded sample file is correct");
+          this.toast.showErrorToast("Ensure uploaded sample files are correct, or try again later.");
```

Comment:
In the case that one or more uploaded samples failed the sketching and AMR process, the user would see this message, but there would be no other indication of what samples had failed, other than that they would be missing from the list displayed - is that correct? I think it would be good for the user to see a list of files which failed in this case, as it's going to be tough for them to work it out if they have hundreds of files. 

Also, I think it's the case that if one sample in batch fails, the remaining successful samples from the batch will not be added the project (since the worker won't get as far as sending the message with the samples array) - I think it would be good if the worker could be more defensive and output both successful and failed samples so successful samples can be included. Again, could be in another ticket!

Reply:
this case ive found never really happens.. I've only noticed it when there is an intermittent error with my machine / web workers... thus I've just added a generic message.. but il make ticket regardless as your point is 100% correct